File: OneCodeTeam/Use-CreateTaskWithContentType-in-SPS-17235/[VB]-Use CreateTaskWithContentType in SPS (VBSharePointTaskWithContentTypeWorkflow)/VB/VBSharePointTaskWithContentTypeWorkflow/MyWorkflow/MyWorkflow.vb
'****************************** Module Header ******************************\
' Module Name:    myWorkflow.vb
' Project:        VBSharePointTaskWithContentTypeWorkflow
' Copyright (c) Microsoft Corporation
'
' The project illustrates how to Use CreateTaskWithContentType in SharePoint Visual Studio Workflow
' 
' This source is subject to the Microsoft Public License.
' See http://www.microsoft.com/en-us/openness/licenses.aspx#MPL.
' All other rights reserved.
' 
' THIS CODE AND INFORMATION IS PROVIDED "AS IS" WITHOUT WARRANTY OF ANY KIND, 
' EITHER EXPRESSED OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE IMPLIED 
' WARRANTIES OF MERCHANTABILITY AND/OR FITNESS FOR A PARTICULAR PURPOSE.
'*****************************************************************************/
Imports System.ComponentModel
Imports System.ComponentModel.Design
Imports System.Workflow.ComponentModel.Compiler
Imports System.Workflow.ComponentModel.Serialization
Imports System.Workflow.ComponentModel
Imports System.Workflow.ComponentModel.Design
Imports System.Workflow.Runtime
Imports System.Workflow.Activities
Imports System.Workflow.Activities.Rules
Imports Microsoft.SharePoint.Workflow
Imports Microsoft.SharePoint.WorkflowActions

Public Class MyWorkflow
    Inherits SequentialWorkflowActivity

    Public Sub New()
        MyBase.New()
        InitializeComponent()
    End Sub

    Public workflowId As Guid = Nothing


    Private Sub createTask1Invoke(sender As Object, e As EventArgs)
        task1Properties = New SPWorkflowTaskProperties()
        task1Properties.Title = String.Format("Please approval:【{0}】", workflowProperties.Item.DisplayName)
        task1Guid = Guid.NewGuid()
        task1ContentTypeId = "0x01080100FFbc98c2529347a5886b8d2576b954e3"
        'This is the Content type which is used by the workFlow
    End Sub

    'ContentTypeId of CreateTaskWithContentType
    Public Shared task1ContentTypeIdProperty As DependencyProperty = DependencyProperty.Register("task1ContentTypeId", GetType(System.String), GetType(MyWorkflow))

    Public Property task1ContentTypeId() As [String]
        Get
            Return DirectCast(MyBase.GetValue(MyWorkflow.task1ContentTypeIdProperty), String)
        End Get
        Set(value As [String])
            MyBase.SetValue(MyWorkflow.task1ContentTypeIdProperty, value)
        End Set
    End Property

    'TaskId of CreateTaskWithContentType
    Public Shared task1GuidProperty As DependencyProperty = DependencyProperty.Register("task1Guid", GetType(System.Guid), GetType(MyWorkflow))

    Public Property task1Guid() As Guid
        Get
            Return CType(MyBase.GetValue(MyWorkflow.task1GuidProperty), System.Guid)
        End Get
        Set(value As Guid)
            MyBase.SetValue(MyWorkflow.task1GuidProperty, value)
        End Set
    End Property

    'TaskProperties of CreateTaskWithContentType
    Public Shared task1PropertiesProperty As DependencyProperty = DependencyProperty.Register("task1Properties", GetType(Microsoft.SharePoint.Workflow.SPWorkflowTaskProperties), GetType(MyWorkflow))

    Public Property task1Properties() As SPWorkflowTaskProperties
        Get
            Return DirectCast(MyBase.GetValue(MyWorkflow.task1PropertiesProperty), Microsoft.SharePoint.Workflow.SPWorkflowTaskProperties)
        End Get
        Set(value As SPWorkflowTaskProperties)
            MyBase.SetValue(MyWorkflow.task1PropertiesProperty, value)
        End Set
    End Property

    'AfterProperties of OnTaskChanged
    Public Shared task1Changed1_AfterPropertiesProperty As DependencyProperty = DependencyProperty.Register("task1Changed1_AfterProperties", GetType(Microsoft.SharePoint.Workflow.SPWorkflowTaskProperties), GetType(MyWorkflow))

    Public Property task1Changed1_AfterProperties() As SPWorkflowTaskProperties
        Get
            Return DirectCast(MyBase.GetValue(MyWorkflow.task1Changed1_AfterPropertiesProperty), Microsoft.SharePoint.Workflow.SPWorkflowTaskProperties)
        End Get
        Set(value As SPWorkflowTaskProperties)
            MyBase.SetValue(MyWorkflow.task1Changed1_AfterPropertiesProperty, value)
        End Set
    End Property

    'BeforeProperties of OnTaskChanged
    Public Shared task1Changed1_BeforePropertiesProperty As DependencyProperty = DependencyProperty.Register("task1Changed1_BeforeProperties", GetType(Microsoft.SharePoint.Workflow.SPWorkflowTaskProperties), GetType(MyWorkflow))

    Public Property task1Changed1_BeforeProperties() As SPWorkflowTaskProperties
        Get
            Return DirectCast(MyBase.GetValue(MyWorkflow.task1Changed1_BeforePropertiesProperty), Microsoft.SharePoint.Workflow.SPWorkflowTaskProperties)
        End Get
        Set(value As SPWorkflowTaskProperties)
            MyBase.SetValue(MyWorkflow.task1Changed1_BeforePropertiesProperty, value)
        End Set
    End Property

    Private isFinished As Boolean = False

    'Handler of OnTaskChanged
    Private Sub task1Changed1Invoke(sender As Object, e As ExternalDataEventArgs)
        'isFinished = task1Changed1_AfterProperties.PercentComplete = 1.0F
        isFinished = ((task1Changed1_AfterProperties.ExtendedProperties(SPBuiltInFieldId.TaskStatus).ToString()) = "Completed")
    End Sub

    Private Sub while1Invoke(sender As Object, e As ConditionalEventArgs)
        e.Result = Not isFinished
    End Sub

    'TaskOutcome of CompleteTask
    Public Shared task1OutcomeProperty As DependencyProperty = DependencyProperty.Register("task1Outcome", GetType(System.String), GetType(MyWorkflow))

    Public Property task1Outcome() As [String]
        Get
            Return DirectCast(MyBase.GetValue(MyWorkflow.task1OutcomeProperty), String)
        End Get
        Set(value As [String])
            MyBase.SetValue(MyWorkflow.task1OutcomeProperty, value)
        End Set
    End Property
    Public Shared workflowPropertiesProperty As DependencyProperty = DependencyProperty.Register("workflowProperties", GetType(Microsoft.SharePoint.Workflow.SPWorkflowActivationProperties), GetType(VBSharePointTaskWithContentTypeWorkflow.MyWorkflow))

    <DesignerSerializationVisibilityAttribute(DesignerSerializationVisibility.Visible)> _
        <BrowsableAttribute(True)> _
            <CategoryAttribute("Misc")> _
    Public Property workflowProperties As SPWorkflowActivationProperties
        Get
            Return CType(MyBase.GetValue(VBSharePointTaskWithContentTypeWorkflow.MyWorkflow.workflowPropertiesProperty), Microsoft.SharePoint.Workflow.SPWorkflowActivationProperties)

        End Get
        Set(value As SPWorkflowActivationProperties)
            MyBase.SetValue(VBSharePointTaskWithContentTypeWorkflow.MyWorkflow.workflowPropertiesProperty, value)

        End Set
    End Property
End Class

File: OneCodeTeam/Use-CreateTaskWithContentType-in-SPS-17236/[C#]-Use CreateTaskWithContentType in SPS (CSSharePointTaskWithContentTypeWorkflow)/C#/CSSharePointTaskWithContentTypeWorkflow/MyWorkflow/MyWorkflow.cs
/****************************** Module Header ******************************\
* Module Name:    myWorkflow.cs
* Project:        CSSharePointTaskWithContentTypeWorkflow
* Copyright (c) Microsoft Corporation
*
* The project illustrates how to Use CreateTaskWithContentType in SharePoint Visual Studio Workflow
* 
* This source is subject to the Microsoft Public License.
* See http://www.microsoft.com/en-us/openness/licenses.aspx#MPL.
* All other rights reserved.
* 
* THIS CODE AND INFORMATION IS PROVIDED "AS IS" WITHOUT WARRANTY OF ANY KIND, 
* EITHER EXPRESSED OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE IMPLIED 
* WARRANTIES OF MERCHANTABILITY AND/OR FITNESS FOR A PARTICULAR PURPOSE.
\*****************************************************************************/

using System;
using System.ComponentModel;
using System.ComponentModel.Design;
using System.Collections;
using System.Workflow.ComponentModel.Compiler;
using System.Workflow.ComponentModel.Serialization;
using System.Workflow.ComponentModel;
using System.Workflow.ComponentModel.Design;
using System.Workflow.Runtime;
using System.Workflow.Activities;
using System.Workflow.Activities.Rules;
using Microsoft.SharePoint;
using Microsoft.SharePoint.Workflow;

namespace MyWorkflow.Workflow
{
    public sealed partial class myWorkflow : SequentialWorkflowActivity
    {
        public myWorkflow()
        {
            InitializeComponent();
        }

        public Guid workflowId = default(System.Guid);
        public SPWorkflowActivationProperties workflowProperties = new SPWorkflowActivationProperties();

        private void createTask1Invoke(object sender, EventArgs e)
        {
            task1Properties = new SPWorkflowTaskProperties();
            task1Properties.Title = string.Format("Please approval:【{0}】", this.workflowProperties.Item.DisplayName);
            task1Guid = Guid.NewGuid();
            //This is the Content type which is used by the workFlow
            task1ContentTypeId = "0x01080100FFbc98c2529347a5886b8d2576b954e3";
        }

        //ContentTypeId of CreateTaskWithContentType
        public static DependencyProperty task1ContentTypeIdProperty = DependencyProperty.Register("task1ContentTypeId", typeof(System.String), typeof(MyWorkflow.Workflow.myWorkflow));

        public String task1ContentTypeId
        {
            get
            {
                return ((string)(base.GetValue(MyWorkflow.Workflow.myWorkflow.task1ContentTypeIdProperty)));
            }
            set
            {
                base.SetValue(MyWorkflow.Workflow.myWorkflow.task1ContentTypeIdProperty, value);
            }
        }

        //TaskId of CreateTaskWithContentType
        public static DependencyProperty task1GuidProperty = DependencyProperty.Register("task1Guid", typeof(System.Guid), typeof(MyWorkflow.Workflow.myWorkflow));

        public Guid task1Guid
        {
            get
            {
                return ((System.Guid)(base.GetValue(MyWorkflow.Workflow.myWorkflow.task1GuidProperty)));
            }
            set
            {
                base.SetValue(MyWorkflow.Workflow.myWorkflow.task1GuidProperty, value);
            }
        }

        //TaskProperties of CreateTaskWithContentType
        public static DependencyProperty task1PropertiesProperty = DependencyProperty.Register("task1Properties", typeof(Microsoft.SharePoint.Workflow.SPWorkflowTaskProperties), typeof(MyWorkflow.Workflow.myWorkflow));

        public SPWorkflowTaskProperties task1Properties
        {
            get
            {
                return ((Microsoft.SharePoint.Workflow.SPWorkflowTaskProperties)(base.GetValue(MyWorkflow.Workflow.myWorkflow.task1PropertiesProperty)));
            }
            set
            {
                base.SetValue(MyWorkflow.Workflow.myWorkflow.task1PropertiesProperty, value);
            }
        }

        //AfterProperties of OnTaskChanged
        public static DependencyProperty task1Changed1_AfterPropertiesProperty = DependencyProperty.Register("task1Changed1_AfterProperties", typeof(Microsoft.SharePoint.Workflow.SPWorkflowTaskProperties), typeof(MyWorkflow.Workflow.myWorkflow));

        public SPWorkflowTaskProperties task1Changed1_AfterProperties
        {
            get
            {
                return ((Microsoft.SharePoint.Workflow.SPWorkflowTaskProperties)(base.GetValue(MyWorkflow.Workflow.myWorkflow.task1Changed1_AfterPropertiesProperty)));
            }
            set
            {
                base.SetValue(MyWorkflow.Workflow.myWorkflow.task1Changed1_AfterPropertiesProperty, value);
            }
        }

        //BeforeProperties of OnTaskChanged
        public static DependencyProperty task1Changed1_BeforePropertiesProperty = DependencyProperty.Register("task1Changed1_BeforeProperties", typeof(Microsoft.SharePoint.Workflow.SPWorkflowTaskProperties), typeof(MyWorkflow.Workflow.myWorkflow));

        public SPWorkflowTaskProperties task1Changed1_BeforeProperties
        {
            get
            {
                return ((Microsoft.SharePoint.Workflow.SPWorkflowTaskProperties)(base.GetValue(MyWorkflow.Workflow.myWorkflow.task1Changed1_BeforePropertiesProperty)));
            }
            set
            {
                base.SetValue(MyWorkflow.Workflow.myWorkflow.task1Changed1_BeforePropertiesProperty, value);
            }
        }

        private bool isFinished = false;

        //Handler of OnTaskChanged
        private void task1Changed1Invoke(object sender, ExternalDataEventArgs e)
        {
            //isFinished = task1Changed1_AfterProperties.PercentComplete == 1F;
            isFinished = (task1Changed1_AfterProperties.ExtendedProperties[SPBuiltInFieldId.TaskStatus].ToString() == "Completed");         
        }

        private void while1Invoke(object sender, ConditionalEventArgs e)
        {
            e.Result = !isFinished;
        }

        //TaskOutcome of CompleteTask
        public static DependencyProperty task1OutcomeProperty = DependencyProperty.Register("task1Outcome", typeof(System.String), typeof(MyWorkflow.Workflow.myWorkflow));

        public String task1Outcome
        {
            get
            {
                return ((string)(base.GetValue(MyWorkflow.Workflow.myWorkflow.task1OutcomeProperty)));
            }
            set
            {
                base.SetValue(MyWorkflow.Workflow.myWorkflow.task1OutcomeProperty, value);
            }
        }

    }
}
